### PR TITLE
unix: use oldest ctime,atime,mtime as birthtime

### PR DESCRIPTION
**Pull request by brendanashworth**
When the operating system does not support birthtime for stat,
previously behaviour was to use the ctime (last changed) time instead.

This change instead enforces that uv_fs_stat will use the oldest
available time (of ctime, atime, and mtime) rather than arbitrarily
picking ctime.

Fixes: https://github.com/nodejs/node/issues/2222

Please note I don't have access to an Android machine to test on.

### CODE REVIEW

File: src/unix/fs.c [hunk 1]
```diff
@@ -701,8 +701,24 @@ static void uv__to_stat(struct stat* src, uv_stat_t* dst) {
   dst->st_mtim.tv_nsec = src->st_mtime_nsec;
   dst->st_ctim.tv_sec = src->st_ctime;
   dst->st_ctim.tv_nsec = src->st_ctime_nsec;
-  dst->st_birthtim.tv_sec = src->st_ctime;
-  dst->st_birthtim.tv_nsec = src->st_ctime_nsec;
+
+  /* On systems that do not support birthtime, resolve the value to the least
+   * time available (oldest of atime, mtime, ctime), disregarding nanoseconds.
+   */
+  if (src->st_atime < src->st_mtime && src->st_atime < src->st_ctime) {
```
Comment (ronkorving) [inline src/unix/fs.c]: Are you not putting the least common case first here? Would the preferred test order not be: ctime, mtime, atime?

Comment (reqshark) [inline src/unix/fs.c]: yea like having a likely/unlikely scenario? though I imagine the compiler doesn't care and will evaluate all branches the same...  http://kernelnewbies.org/FAQ/LikelyUnlikely

Comment (brendanashworth) [inline src/unix/fs.c]: The actual comparison order is mostly arbitrary, but all cases are (I think, cause of `&&`) equal in comparison time at 2 checks, except for the worst case (ctime < atime < mtime). The worst case is also arbitrary, but I can change that if we think there is a better one.

Comment (ronkorving) [inline src/unix/fs.c]: I'm just thinking of the most likely true case to be the first one in the list. Good for branch prediction (perhaps?) and just overall avoiding needless comparisons. Just my 2 cents, do as you please of course :)

**Comment by saghul** [comment]
I'm not sure I like this change. Even if currently incorrect on platforms without support for st_birthtime, we have always kept it to be the same as ctime: https://github.com/libuv/libuv/commit/f9e6029b8215772b8c20d3fee85835dfb101dc57 (you can blame yours trully for that). So changing it to another value which is not correct either is not that much of an improvement, IMHO.

Some filesystems (ext4 for instance) seem to record the file creation time, but I couldn't find the way to get that without parsing the output of `debugfs` (I just had a quick look).

On hindsight, I should have zeroed that field out, oh well!

**Comment by brendanashworth** [comment]
Would this be something you'd like zero'd now? Or would that break backwards compatibility?

It seems to me that if libuv will give a value for birthtime, it might as well give the oldest value it has, rather than pinning it at ctime. I can see it being very confusing for a developer that expects birthtime to be `<= ctime,atime,mtime`, which is a sound expectation on all operating systems minus ones that support don't birthtime. It's less incorrect, in my opinion, though not perfect.

**Comment by saghul** [comment]
> Would this be something you'd like zero'd now? Or would that break backwards compatibility?

I would like to have it zeroed, but I'd say that would break backwards compatibility.

@bnoordhuis any thoughts on this?

**Comment by bnoordhuis** [comment]
I agree, it breaks backwards compatibility.

**Comment by saghul** [comment]
Sorry @brendanashworth I'm closing this. A note in the documentation pointing out this caveat would be a welcome PR though.

**Comment by brendanashworth** [comment]
Thanks for taking a look, I'll send in a PR sometime soon.

**Pull request closed without merging**